### PR TITLE
Add mu4e layer

### PR DESCRIPTION
**Pull request by darkfeline**
Branch for developing mu4e layer for Spacemacs.

### CODE REVIEW

**Comment by numkem** [comment]
I can vouch that it works without problems.

**Comment by myrjola** [comment]
Nice! :+1: Looking forward to get this merged.

**Comment by gabrielalmeida** [comment]
Great! Note that it breaks ```contrib/spotify``` which does use ```SPC m``` as ```group:music```, bind should be remapped.

**Comment by darkfeline** [comment]
@gabrielalmeida Yuck.  I think an email client should take priority over using Emacs to control Spotify.  Let Spotify be `SPC a M`.  Worst case I'll make mu4e take `SPC a M`, but that disgusts me a little.

**Comment by andreas-h** [comment]
+1

**Comment by syl20bnr** [comment]
> Let Spotify be `SPC a M`

:+1:

**Comment by cmccloud** [comment]
If there is any interest in adding some optional g-mail specific functionality to this layer, you can take a look at my config here: https://github.com/cmccloud/.spacemacs.d/blob/master/layers/mu4e/extensions.el

**Comment by darkfeline** [comment]
@cmccloud I will look into borrowing some of your code, or you can pull request if you want direct credit for it.

Just the extra evilify bindings for now.  I don't need the extra Gmail features myself since I keep email simple (read, delete, save a copy in my org-mode notes if I need to keep it).  If others want it then I will take it if they don't interfere with non-Gmail users.

Why do you use progn in use-package?  `:init` and `:config` can take multiple expressions.

**Comment by syl20bnr** [comment]
Spacemacs uses `progn` also ;-)

more info here: https://github.com/syl20bnr/spacemacs/pull/2255

I will add it to the conventions.

**Comment by syl20bnr** [comment]
Also I'm going to merge this soon to @cmccloud can add some config if he wants.

**Comment by cmccloud** [comment]
@darkfeline feel free to take whatever you like - I thought that some of the gmail code might be useful as examples in the documentation since it's such a popular service.

**Comment by cmccloud** [comment]
@darkfeline Feel free to borrow whatever you like. Honestly I think the best place for a lot of that code might be in the documentation as an example of how to add gmail specific functionality. On the other hand, none of the code will interfere with other mail services, but it may have funny behavior. 

e.g. adding a /Trash tag instructs g-mail to move a message to the trash - most other services probably wont do that.

**Comment by cmccloud** [comment]
@darkfeline i've refactored my mu4e config to use the new packages.el features if you're interested in taking a look.
@syl20bnr please correct me if i've used these incorrectly, but they seem to work just fine

https://github.com/cmccloud/.spacemacs.d/blob/master/layers/mu4e/packages.el

**Comment by syl20bnr** [comment]
I highly suggest to suppress `:step post`, actually I may remove this keyword because it will harm more than it helps. Only the pre step is really useful for some important package that provide services useful for the other packages.

The packages are loaded in alphabetical order so the loading order is deterministic and easy to know in advance. So if A require B then we can use some `eval-after-load` black magic. Throwing post step in this will just remove the benefit of the deterministic loading behavior.

**Comment by boj** [comment]
:+1:

**Comment by nwolfe** [comment]
Ping @darkfeline

**Comment by cpaulik** [comment]
:+1: 

When this is merged I can add a few nice features that I have collected. Like attaching files using helm.

File: contrib/!email/mu4e/funcs.el [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+                (match-string 1 maildir))
+            (helm-comp-read
+             "Compose with account:"
+             (mapcar (lambda (var) (car var)) mu4e-account-alist))))
+         (account-vars (cdr (assoc account mu4e-account-alist))))
+    (if account-vars
+        (mu4e//map-set account-vars)
+      (error "No email account found"))))
+
+(defun mu4e/msgv-action-view-in-browser (msg)
```
Comment (nwolfe) [inline contrib/!email/mu4e/funcs.el]: There's a built-in function called `mu4e-action-view-in-browser` that does this

**Comment by darkfeline** [comment]
This should be mergeable as is.  I'm a bit hesitant to add anything else since I only use basic email functionality, and I don't want to push any code I haven't tested personally.

@syl20bnr Please merge unless you have additional comments, the community can add more features as they see fit.

**Comment by nwolfe** [comment]
@darkfeline It seems like @syl20bnr was waiting for you to remove the `:step post` line?  That's how I'm reading his comment at least :smiley: 

Also, the directory structure for layers has changed since this PR was submitted.  This layer should now be under `layers/+email/mu4e`.

This is awesome, looking forward to getting this merged in so the community can build this out further.

**Comment by darkfeline** [comment]
@nwolfe 

>Also, the directory structure for layers has changed since this PR was submitted. This layer should now be under layers/+email/mu4e.

Duh, I forgot about that.

>It seems like @syl20bnr was waiting for you to remove the :step post line? That's how I'm reading his comment at least ￼

I don't think I've used `:step post` anywhere.

(As an aside/warning, I did a `rebase/push --force`, which is probably not recommended Git usage.  The alternative would be to merge develop -> mu4e, but I didn't know if Github would then decide to include all 500+ of those commits into the PR as well.  It looks like Github handled the forced rebase fairly cleanly?)

**Comment by nwolfe** [comment]
@darkfeline Awesome, thanks for updating.  Yeah I don't know what the `:step post` stuff was about.

Shoot, I forgot to mention that @syl20bnr prefers PRs to be a single commit.  Could you please squash these commits into one?  That should be the last hurdle for merging/cherry-picking this in!

Your git moves are good; `rebase` is necessary and correct, and you'll have to `push --force` once you squash the commits locally.

**Comment by darkfeline** [comment]
@nwolfe Done.

**Comment by nwolfe** [comment]
:+1:

**Comment by nwolfe** [comment]
Bump - I believe this is ready for merge

**Comment by dsdshcym** [comment]
:+1: for this layer. mu4e is way better than gnus.

**Comment by nwolfe** [comment]
@TheBB @syl20bnr Please merge

**Comment by TheBB** [comment]
Maybe this is a stupid question, but where is `mu4e` itself? It's listed in this PR as an extension, but the code for it is not included. I also can't find a package called `mu4e` on MELPA.

**Comment by cpaulik** [comment]
`mu4e` is installed when you install `mu`. Normally into `/usr/share/emacs/site-lisp`. The site-lisp path should normally be in the `load-path`.

**Comment by nwolfe** [comment]
> The site-lisp path should normally be in the load-path.

That's an exercise left for the reader, right? Not something we can hard-code into this layer itself?

**Comment by cpaulik** [comment]
I don't know if the path to `site-lisp` is the same for all linux distributions and I have no idea if it is fixed on OSX or Windows (if mu works on those systems?)

**Comment by mclearc** [comment]
For OS X it is going to depend on whether the user is using stock emacs or emacs from hombrew. The latter has `site-lisp` in `usr/local/share/emacs/`

**Comment by nwolfe** [comment]
Yeah, not something we can hardcode then.  Probably worthwhile to document in the README once this is merged though.

**Comment by d12frosted** [comment]
But then you should add a note in README about it.

**Comment by numkem** [comment]
I've tried using this layer again and I still can't get the detection of account to work properly. when replying or sending new email it always reverts to `<username>@<hostname> as the `From` address.

I've followed the `README` and started from scratch.

**Comment by tsenart** [comment]
+1

**Comment by declanqian** [comment]
@darkfeline Please add some text about mu4e package as @d12frosted mentioned before. So this can be merged? Awesome work by the way

**Comment by TheBB** [comment]
Thank you! Cherry-picked in develop. You can safely delete your branch. :tada:

I made some changes in https://github.com/syl20bnr/spacemacs/commit/5f3793a50e9be0b500891088a79f87523f2c167f:

- Clarified installation instructions (as per above discussion) and bring the README in line with the current documentation conventions.
- Added a layer variable for installation path instead of having the user mess with `load-path` in the init phase.
- Moved `mu4e-account-alist` to `config.el` where it belongs.
- Moved the main code to `packages.el` since `extensions.el` is now deprecated.
- Define the `mu4e` package as built-in and not as local.
- Auto-load `mu4e` on invocation of the two given keybindings, since prior to this it was not deferred.

I should note that I haven't used this layer so please, those of you who intend to use it let me know if I have messed something up. I will probably try it out in the future. :-)

**Comment by d12frosted** [comment]
Nice, mu4e is now part of Spacemacs. Going to try it.

**Pull request closed without merging**